File: Parameters.fs
module Parameters

open System
open FSharpTools
open FSharpTools.Functional

type Mode =
| Create
| Operate
| Delete

type Value = {
    Staging: bool
    Mode: Mode
}

/// Evaluating command args
/// 
/// the following arguments are valid:
/// 
/// * ```-prod``` productive, without: staging (test)
/// * ```-create``` creates file cert.json, then exits. Setup step
/// * ```-del``` delete account
/// 
///**Returns**
/// ```Value``` object
let get = 
    let get () = 
        
        let printProductive b =
            if b then "!!! P R O D U C T I V E !!!" else "Staging" 
            |> printfn "%s"

        let args = Environment.GetCommandLineArgs ()
        {
            Staging = 
                args 
                |> Array.contains "-prod" 
                |> sideEffect printProductive 
                |> not
            Mode = 
                match args |> Array.contains "-create", args |> Array.contains "-del" with
                | true, false -> Create
                | false, true -> Delete
                | _           -> Operate
        }
    memoizeSingle get

let getEncryptDirectory =
    let getEncryptDirectory () =
        Environment.GetFolderPath Environment.SpecialFolder.ApplicationData
        |> Directory.attachSubPath "letsencrypt-uweb"
    memoizeSingle getEncryptDirectory

let getCertFile = 
    let getCertFile () = 
        getEncryptDirectory ()
        |> Directory.attachSubPath "cert.json"
    memoizeSingle getCertFile
    
let getAccountFile = 
    let getAccountFile () =
        getEncryptDirectory ()
        |> Directory.attachSubPath "accounts.pem"

    memoizeSingle getAccountFile

File: Parameters.cs
using Certes.Acme;
using CsTools.Extensions;

using static System.Console;
using static CsTools.Functional.Memoization;

record Parameters(
    bool Staging,
    OperationMode Mode
) {
    public static Func<Parameters> Get { get; }
        = Memoize(Init);

    public static Func<string> GetEncryptDirectory { get; }
        = Memoize(InitEncryptDirectory);

    public static Func<string> GetPfxFile { get; }
        = Memoize(InitGetPfxFile);

    public static Func<string?> GetCertFile { get; }
        = MemoizeMaybe(InitGetCertFile);

    public static Func<string> GetAccountFile { get; }
        = Memoize(InitGetAccountFile);

    public static Func<string> GetPfxPassword { get; }
        = Memoize(InitGetPfxPassword);

    public static Func<Uri> GetAcmeUri { get; }
        = Memoize(InitGetAcmeUri);

    static Parameters Init()
        => Environment.GetCommandLineArgs()
            .Pipe(args => new Parameters(
                args
                    .Contains("-prod")
                    .SideEffect(n => WriteLine(n ? "!!! P R O D U C T I V E !!!" : "Staging"))
                    != true,
                (args.Contains("-create"), args.Contains("-del")) switch
                {
                    (true, false) => OperationMode.Create,
                    (false, true) => OperationMode.Delete,
                    _ => OperationMode.Operate
                })
                    .SideEffect(WriteLine)
            );

    static string InitEncryptDirectory()
        => Environment
            .GetFolderPath(Environment.SpecialFolder.ApplicationData)
            .AppendPath("letsencrypt-uweb");

    static string InitGetPfxFile()
        => GetEncryptDirectory()
            .AppendPath(Get().Staging ? "certificate-staging.pfx" : "certificate.pfx");

    static string? InitGetCertFile()
        => GetEncryptDirectory()
            .AppendPath("cert.json")
            .CheckIfFileExists();
            

    static string InitGetAccountFile()
        => GetEncryptDirectory()
            .AppendPath(Get().Staging ? "account-staging.pem" : "account.pem");

    static string InitGetPfxPassword()
        => Get()
            .Staging
                ? "/etc"
                : Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData)
            .AppendPath("LetsencryptUweb")
            .ReadAllTextFromFilePath()
            ?.Trim()
            ?? "".SideEffect(_ => WriteLine("!!!NO PASSWORD!!"));

    static Uri InitGetAcmeUri()
        => Get().Staging
            ? WellKnownServers.LetsEncryptStagingV2
            : WellKnownServers.LetsEncryptV2;
}